<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        /* 参考:[太复杂了!!!]
        掘金:https://juejin.cn/post/6844904096525189128?searchId=2023080319322662CA6BB2944F1B0B7C7E */

        /* promise调用流程? */
        const p1 = new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve('result')
            }, 1000);
        })

        p1.then(res => console.log(res), err => console.log(err))

        /* 原生promise实现? */
        class MyPromise {
            // 构造方法接收一个回调
            constructor(executor) {
                this._resolveQueue = []    // then收集的执行成功的回调队列
                this._rejectQueue = []     // then收集的执行失败的回调队列

                // 由于resolve/reject是在executor内部被调用, 因此需要使用箭头函数固定this指向, 否则找不到this._resolveQueue
                let _resolve = (val) => {
                    // 从成功队列里取出回调依次执行
                    while (this._resolveQueue.length) {
                        const callback = this._resolveQueue.shift()
                        callback(val)
                    }
                }
                // 实现同resolve
                let _reject = (val) => {
                    while (this._rejectQueue.length) {
                        const callback = this._rejectQueue.shift()
                        callback(val)
                    }
                }
                // new Promise()时立即执行executor,并传入resolve和reject
                executor(_resolve, _reject)
            }

            // then方法,接收一个成功的回调和一个失败的回调，并push进对应队列
            then(resolveFn, rejectFn) {
                this._resolveQueue.push(resolveFn)
                this._rejectQueue.push(rejectFn)
            }
        }
        const p2 = new MyPromise((resolve, reject) => {
            setTimeout(() => {
                resolve('成功')
            }, 1000)
        })
        p2.then(res => console.log(res)) /* 不能链式调用.then(res => console.log('链式then', res)) */

        /* Promise A+规范[要有状态]
        1.Promise本质是一个状态机，且状态只能为以下三种：Pending（等待态）、Fulfilled（执行态）、Rejected（拒绝态），状态的变更是单向的，只能从Pending -> Fulfilled 或 Pending -> Rejected，状态变更不可逆
        2.then方法接收两个可选参数，分别对应状态改变时触发的回调。then方法返回一个promise。then 方法可以被同一个 promise 调用多次。
        */

        //Promise/A+规范的三种状态
        const PENDING = 'pending'
        const FULFILLED = 'fulfilled'
        const REJECTED = 'rejected'

        class MyApPromise {
            constructor(executor) {
                this._status = PENDING     // Promise状态
                this._resolveQueue = []    // 成功队列, resolve时触发
                this._rejectQueue = []     // 失败队列, reject时触发

                let _resolve = (val) => {
                    if (this._status !== PENDING) return   // 对应规范中的"状态只能由pending到fulfilled或rejected"
                    this._status = FULFILLED              // 变更状态

                    // 队列存储回调,因为"then 方法可以被同一个 promise 调用多次"
                    while (this._resolveQueue.length) {
                        const callback = this._resolveQueue.shift()
                        callback(val)
                    }
                }
                // 实现同resolve
                let _reject = (val) => {
                    if (this._status !== PENDING) return   // 对应规范中的"状态只能由pending到fulfilled或rejected"
                    this._status = REJECTED               // 变更状态
                    while (this._rejectQueue.length) {
                        const callback = this._rejectQueue.shift()
                        callback(val)
                    }
                }
                // new Promise()时立即执行executor,并传入resolve和reject
                executor(_resolve, _reject)
            }

            // then方法,接收一个成功的回调和一个失败的回调
            // then(resolveFn, rejectFn) {
            //     this._resolveQueue.push(resolveFn)
            //     this._rejectQueue.push(rejectFn)
            // }
            // then方法
            then(resolveFn, rejectFn) {
                //return一个新的promise
                return new MyApPromise((resolve, reject) => {
                    //把resolveFn重新包装一下,再push进resolve执行队列,这是为了能够获取回调的返回值进行分类讨论
                    const fulfilledFn = value => {
                        try {
                            //执行第一个(当前的)Promise的成功回调,并获取返回值
                            let x = resolveFn(value)
                            //分类讨论返回值,如果是Promise,那么等待Promise状态变更,否则直接resolve
                            //这里resolve之后，就能被下一个.then()的回调获取到返回值，从而实现链式调用
                            x instanceof MyApPromise ? x.then(resolve, reject) : resolve(x)
                        } catch (error) {
                            reject(error)
                        }
                    }
                    //把后续then收集的依赖都push进当前Promise的成功回调队列中(_rejectQueue), 这是为了保证顺序调用
                    this._resolveQueue.push(fulfilledFn)

                    //reject同理
                    const rejectedFn = error => {
                        try {
                            let x = rejectFn(error)
                            x instanceof MyApPromise ? x.then(resolve, reject) : resolve(x)
                        } catch (error) {
                            reject(error)
                        }
                    }
                    this._rejectQueue.push(rejectedFn)
                })
            }

        }

        /* 原生promise */
        const p3 = new Promise((resolve, reject) => {
            resolve(1)
        })

        p3
            .then(res => {
                console.log(res)
                return new Promise((resolve, reject) => {
                    setTimeout(() => {
                        resolve(2)
                    }, 1000);
                })
            })
            .then(res => {
                console.log(res)
                return 3
            })
            .then(res => {
                console.log(res)
            })

        /* 实现队列+状态+链式then后的promise */
        const p4 = new MyApPromise((resolve, reject) => {
            resolve(1)
        })

        p4
            .then(res => {
                console.log(res)
                return new Promise((resolve, reject) => {
                    setTimeout(() => {
                        resolve(2)
                    }, 1000);
                })
            })
            .then(res => {
                console.log(res)
                return 3
            })
            .then(res => {
                console.log(res)
            })

    </script>
</body>

</html>